File: rust/src/0023-merge_k_sorted_lists.rs
pub use std::collections::BinaryHeap;

pub struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl PartialOrd for ListNode {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for ListNode {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.val.cmp(&other.val).reverse()
    }
}

impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        let mut heap = lists
            .into_iter()
            .filter_map(|x| x)
            .collect::<BinaryHeap<_>>();
        let mut head = None;
        let mut tail = &mut head;
        while let Some(node) = heap.pop() {
            tail = &mut tail.insert(node).next;
            if let Some(next) = tail.take() {
                heap.push(next)
            }
        }
        return head;
    }
}

File: cpp/0023-merge_k_sorted_lists.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

using namespace std;

class Solution {
public:
  ListNode *mergeKLists(vector<ListNode *> &lists) {
    // c++ is fucked up man...
    auto cmp = [](const ListNode *l1, const ListNode *l2) {
      return l1->val > l2->val;
    };
    priority_queue<ListNode *, vector<ListNode *>, decltype(cmp)> heap(cmp);
    for (int i = 0; i < lists.size(); i++) {
      if (lists[i] != nullptr) {
        heap.push(lists[i]);
        lists[i] = lists[i]->next;
      }
    }
    ListNode *head = nullptr;
    ListNode **ass = &head;
    while (!heap.empty()) {
      *ass = heap.top();
      heap.pop();
      if ((*ass)->next != nullptr) {
        heap.push((*ass)->next);
      }
      ass = &(*ass)->next;
    }
    return head;
  }
};
